### PR TITLE
hack sys.argv to fix #2389 on macos

### LINKED ISSUE #2389: mac thinks that napari is named "__main__.py" in bundled app
## 🐛 Bug

My mac seems to think that napari is named "__main__.py" in bundled app

## To Reproduce

1. Launch napari bundled app on mac
2. Click "napari" on menu bar
3. See the menu items "Hide __main__.py" and "Close __main__.py"

image

## Expected behavior

menu items include "Hide napari" and "Close napari", similar to Chrome...

image

## Environment

napari: 0.4.6
Platform: macOS-10.15.7-x86_64-i386-64bit
System: MacOS 10.15.7
Python: 3.8.7 (default, Jan 2 2021, 04:16:43) [Clang 11.0.0 (clang-1100.0.33.17)]
Qt: 5.15.2
PySide2: 5.15.2
NumPy: 1.19.3
SciPy: 1.6.1
Dask: 2021.03.0
VisPy: 0.6.6

OpenGL:
- GL version: 2.1 ATI-3.10.19
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1792x1120, scale 2.0
- screen 2: resolution 2560x1440, scale 1.0

Plugins:
- affinder: 0.1.0
- aicsimageio: 0.2.0
- aicsimageio_delayed: 0.2.0
- brainreg: 0.2.3
- brainreg_standard: 0.2.3
- cellfinder: 0.2.1
- console: 0.0.3
- napari-hdf5-labels-io: 0.2.dev1
- ndtiffs: 0.1.1
- svg: 0.1.4

### PR DESCRIPTION
**Pull request by jaimergp**
# Description

## Type of change

- [X] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

Closes #2389 

QMenuBar will populate the App menu on MacOS by using the basename of `sys.argv[0]`. We check whether this is not `napari` and, in that case, we add a new a path component to make Qt believe it was (before any qt imports). We undo that change as quickly as possible, after QApplication has been instantiated. 

# How has this been tested?

- [X] Opened a napari installation with `python -m napari` and check name is napari
- [X] sys.argv is restored in IPython console

__BEFORE__

image

__AFTER__

image

## Final checklist:
- [X] My PR is the minimum possible work for the desired functionality
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/__main__.py [hunk 1]
```diff
@@ -441,6 +441,13 @@ def main():
 
         multiprocessing.set_start_method('fork')
 
+    # Prevent __main__.py from being the app name if run with python -m napari
+    if sys.platform == "darwin" and not sys.argv[0].endswith("/napari"):
+        # we need a path with basename `napari`; adding ../ makes it an existing
+        # path too, just in case -- we undo this later once QApplication() is created
+        sys._old_argv_0 = sys.argv[0]
```
Comment (jni) [inline napari/__main__.py]: @jaimergp can you tell me what `sys.argv[0]` *was* in this case? ie why was the basename *not* napari?

Comment (jni) [reply napari/__main__.py]: Or is this `python`? And if so, why replace it with `/../napari`? (ie I don't see why that is guaranteed to exist)

Comment (jaimergp) [reply napari/__main__.py]: If invoked with `python -m napari`, it is set to `/path/to/installation/napari/__main__.py`. See screenshot in PR description, which is what ends up in the application menu.

Comment (jaimergp) [reply napari/__main__.py]: Adding `/../napari` turns `sys.argv[0]` into `/path/to/installation/napari/__main__.py/../napari`, which should be equivalent to `/path/to/installation/napari`.

Comment (jaimergp) [reply napari/__main__.py]: Ah wait, it should be `/../../napari` in that case...

```python
>>> import os
>>> os.path.normpath("/path/to/installation/napari/__main__.py/../napari")
'/path/to/installation/napari/napari'
>>> os.path.normpath("/path/to/installation/napari/__main__.py/../../napari")
'/path/to/installation/napari
```

**Comment by Carreau** [comment]
It looks to me like `sys.argv` is passed explicitly to `QApplication` in `qt_event_loop.py`. Wouldn't it be better to create a copy and pass the copy to QApplication instead of messing with the original ones ?

**Comment by jaimergp** [comment]
Hm, let me look into that @Carreau. Thanks!

**Review by Carreau — APPROVED** [review]
That looks much cleaner, and I'm much more comfortable with this.
I'll let @jni merge or approve as he reviewed as well.

**Comment by Carreau** [comment]
it's strange as the windows-latest python-3.7 test failed twice on the same test with this change. Let me rerun a 3rd time.

**Comment by Carreau** [comment]
> it's strange as the windows-latest python-3.7 test failed twice on the same test with this change. Let me rerun a 3rd time.

Or failure on windows disappeared and the other one seem unrelated. Merging.

**Pull request merged**